Updating SDK to 0.5.2 spec

Norhaven opened the pull request:
This implements the C# SDK to be compliant with the 0.5.2 version of the GrowthBook SDK spec. Things that are a part of this include:
- New unit test structure for easier use of the standard cases.json tests (all tests are currently passing).
- New feature repository/cache with support for Server Sent Events, background retrieval, and encrypted features.
- Extensive logging which can be enabled by log level (or overridden with a custom implementation).
- More robust error handling.

Please let me know if you have any questions.

Thanks!

darklordzw commented:
@Norhaven, looks great, thanks for contributing such as substantial update! My one question is why you changed all of the double's to floats for things experiment weights? I'm not necessarily opposed to it, but I'm curious as to the reason.

darklordzw commented:
@Mastermindzh, @pascal89 any thoughts on this PR before I merge this in?

Mastermindzh commented:
@darklordzw I'll have a look tomorrow morning (CEST) and let you know 😊

Norhaven commented:
@darklordzw The use of floats instead of doubles was primarily because it seemed like the typical floating point values needed by things like weights didn't require more than ~7 digits of precision to be effective, but I'm open to either way if there's a preference for doubles in particular. Thoughts?

pascal89 commented:
@darklordzw I'll schedule a review moment for tomorrow

darklordzw commented:
@Norhaven, gotcha. My original implementation used floats, but I ended up switching to doubles to be consistent with the other Growthbook sdks (this was originally a port of the Python version). I don't have a strong opinion either way, but I lean towards staying with doubles to stay consistent with what's out there already.

Norhaven commented:
@darklordzw Sure, makes sense to me. I've replaced the float usages with doubles now. The precision change caused a minor rounding difference in the GetEqualWeights test that differs from the spec, so I've called that out in there with a little more info and adjusted it slightly. Let me know if you have more questions on anything, thanks!

Mastermindzh commented on GrowthBook.Tests/ApiTests/LoadFeaturesTests.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    private const string ApiKey = "<AN API KEY GOES HERE>";
+
+    [Fact(Skip = "This is here as a shortcut for the purpose of debugging this SDK against the actual API and should not be used otherwise")]
```
I don't like including these things t.b.h

Norhaven replied:
Yeah, I was a little on the fence about that as well. Removed.

Mastermindzh commented on GrowthBook.Tests/ApiTests/LoadFeaturesTests.cs:
```diff
@@ -0,0 +1,32 @@
+
+public class LoadFeaturesTests
+{
+    private const string ApiKey = "<AN API KEY GOES HERE>";
```
If we're including this file I'd recommend using IConfiguration for these things:
See, [this](https://dev.to/andreav/using-dotnet-iconfiguration-inside-xunit-projects-4o6b) for an example.

Norhaven replied:
File is removed now.

Mastermindzh commented on GrowthBook/AssemblyInfo.cs:
```diff
@@ -0,0 +1,6 @@
+using System.Runtime.CompilerServices;
```
I suggest removing this file altogether and stick with the portable configuration in csproj:

```csharp
   <ItemGroup>
    <InternalsVisibleTo Include="GrowthBook.Tests" />
  </ItemGroup>
```

Norhaven replied:
Good idea, forgot about that. Done.

Mastermindzh commented on GrowthBook/FeatureResult.cs:
```diff
@@ -8,6 +9,17 @@ namespace GrowthBook
+        /// <summary>
+        /// Represents possible values for the source of a feature result.
+        /// </summary>
+        public static class SourceIs
```
I would recommend removing since it isn't used.
Also, I think you meant "SourceIDs"

Norhaven replied:
It's actually used when calling GetFeatureResult and was intended to be a more fluent-property kind of thing so it reads FeatureResult.SourceIs.DefaultValue or similar. Do you feel that it's more confusing that way?

Norhaven replied:
Actually, the more I think about it, we don't currently have much of a fluent-syntax usage in this, would add to the confusion. Renamed.

Mastermindzh commented on GrowthBook/GrowthBook.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <Authors>Ben Whatley</Authors>
+    <Authors>Ben Whatley,Norhaven</Authors>
```
I would usually only put recognized contributors of the Github repository here

Norhaven replied:
Fair enough, removed. :)

Mastermindzh commented on GrowthBook/GrowthBook.csproj:
```diff
@@ -11,11 +11,17 @@
     <RepositoryUrl>https://github.com/growthbook/growthbook-csharp.git</RepositoryUrl>
     <RepositoryType>git</RepositoryType>
     <PackageTags>GrowthBook,A/B test,feature toggle</PackageTags>
```
Would recommend to include "flag"

Norhaven replied:
Added now.

Norhaven commented on GrowthBook/Api/FeatureRefreshWorker.cs:
```diff
@@ -0,0 +1,202 @@
+                                    continue;
+                                }
+
+                                json = json.Substring(5).Trim();
```
Hmm... not sure I follow, but added some more comments to hopefully clarify the reasoning behind doing it this way. Let me know if I missed the point. :)

Norhaven commented on GrowthBook/Api/HttpClientFactory.cs:
```diff
@@ -0,0 +1,59 @@
+{
+    public class HttpClientFactory : IHttpClientFactory
+    {
+        public static class ConfiguredClients
```
Good idea, moved it out now.

Norhaven commented on GrowthBook.Tests/GrowthBook.Tests.csproj:
```diff
@@ -31,6 +31,7 @@
     <PackageReference Include="FluentAssertions" Version="6.5.1" />
     <PackageReference Include="Microsoft.AspNetCore.Mvc.Testing" Version="5.0.10" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
+    <PackageReference Include="Moq" Version="4.20.70" />
```
Hmm... I remember that. It looks like they've backed off from that SponsorLink implementation a little with the security issues but I'm open to trying other options if there's still concern about that.

Norhaven commented on GrowthBook/Extensions/ValidationExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+        /// </remarks>
+        /// <param name="value">The string to verify.</param>
+        /// <returns>True if null, empty, or whitespace, false otherwise.</returns>
+        public static bool IsMissing(this string value) => string.IsNullOrWhiteSpace(value);
```
Ahh, gotcha. Was trying to remove some of the boilerplate and focus more on the semantics of what the caller is trying to achieve, but looks like it missed the mark. Will take another look at that, thanks.

Norhaven commented:
@Mastermindzh Thanks for the review, updated the PR and had a few questions.

pascal89 reviewed (commented):
I didn't review the test cases due to a lack of time.
I really like the infrastructure for keeping the features up-to-date (with the server side events) :+1: :+1:. It would be nice to have some usage documentation in the README.md, as any local implementation of this package will probably be impacted because of these changes. Also, I'm not sure how to work with attributes. We are using this package in a web API and for each HTTP request I need to set the attributes in order to target features to a certain audience (user or user role). Could you shed some light on this?, preferably in the form of documentation.

Norhaven commented on GrowthBook/Api/InMemoryFeatureCache.cs:
```diff
@@ -0,0 +1,77 @@
+            }
+        }
+
+        public async Task RefreshWith(IDictionary<string, Feature> features, CancellationToken? cancellationToken = null)
```
I made that call return a Task in the IGrowthBookFeatureCache interface that this implements because, assuming usage is backend in a distributed environment (e.g. cloud), people may want to provide their own cache that works better for their situation, like a Redis-backed one or similar, instead of just an in-memory one. That kind of thing is likely to be more appropriately accessed in an async manner, so I preemptively went that direction. 

And you're right, the InMemoryFeatureCache itself has no async needs, but it should probably conform to that pattern anyway. I made a quick change now to remove the 'async' keyword and just explicitly return a Task.CompletedTask to make things a little more clear and avoid the warning, hopefully that all helps. :)

Norhaven commented on GrowthBook/Experiment.cs:
```diff
@@ -26,7 +26,7 @@ public class Experiment
         /// How to weight traffic between variations. Must add to 1.
         /// </summary>
-        public IList<double> Weights { get; set; }
+        public List<double> Weights { get; set; }
```
I'd agree. Hmm... from my perspective, I would tend to prefer an array in here on data objects just so we can both have explicit length checks/indexing available as well as enforce the readonly semantics a little more to discourage adding/removing items, but with the existing SDK implementation using lists I wasn't sure if there was a use case for people wanting to add their own values on the fly to these objects. Taking a breaking change to that might cause problems. 

@darklordzw Just wondering, did you have a particular design decision in mind choosing lists over something else for the existing implementation? Might help inform the directions we can go here. :)

Norhaven commented on GrowthBook/GrowthBook.cs:
```diff
@@ -151,253 +184,402 @@ public Action Subscribe(Action<Experiment, ExperimentResult> callback)
-        /// <param name="experiment">The experiment to evaluate.</param>
-        /// <returns>The experiment result.</returns>
-        ExperimentResult RunExperiment(Experiment experiment)
+        public async Task LoadFeatures(GrowthBookRetrievalOptions options = null, CancellationToken? cancellationToken = null)
```
Good point, added. :)

Norhaven commented on GrowthBook/Api/FeatureRefreshWorker.cs:
```diff
@@ -0,0 +1,209 @@
+
+            if (_config.PreferServerSentEvents)
+            {
+                _isServerSentEventsEnabled = response.Headers.TryGetValues("x-sse-support", out var values) && values.Contains("enabled");
```
Sounds good, done.

Norhaven commented on GrowthBook/Api/FeatureRefreshWorker.cs:
```diff
@@ -0,0 +1,209 @@
+            var hostEndpoint = config.ApiHost;
+            var trimmedHostEndpoint = new string(hostEndpoint?.Reverse().SkipWhile(x => x == '/').Reverse().ToArray());
+
+            _featuresApiEndpoint = $"{trimmedHostEndpoint}/api/features/{config.ClientKey}";
```
Ahh, gotcha. Will look into that, thanks!

Norhaven commented on GrowthBook/GrowthBook.cs:
```diff
@@ -32,16 +41,45 @@ public GrowthBook(Context context)
+            var featureCache = new InMemoryFeatureCache(cacheExpirationInSeconds: 60);
+            var httpClientFactory = new HttpClientFactory(requestTimeoutInSeconds: 60);
+
+            var loggerFactory = context.DefaultLoggerFactory ?? LoggerFactory.Create(builder =>
```
Yeah, I think you're right. I'd added in the console/debug logging for testing and provided the DefaultLogLevel property on the Context that could modify this, but looking at it some more I think that's an extraneous property and this can be condensed and not force a particular logging provider on the user. Will fix that. Thanks!

Norhaven commented:
@darklordzw @Mastermindzh @pascal89 Okay, the PR is ready for another look over, please let me know if you have questions. I've left the existing comments open for any further discussion necessary, feel free to resolve them if it works for you. Thanks!

Mastermindzh commented:
@Norhaven @darklordzw 

All my feedback has been implemented and then some. I don't see a reason not to merge this 😄.
Great job @Norhaven !

pascal89 commented:
@Norhaven Great work. Still have a little remark, though. It would be nice if you include two additions:
1. Update the `README.md` because setting up GrowthBook in any implementation has changed a bit because of the inftrastructure for fetching the features and keeping it up-to-date is now included in the package.
2. Update the `CHANGELOG.md` with the changes you've made.

Norhaven commented:
@pascal89 Just updated the changelog now, thanks! For the readme and documentation, I also have an extensive ASP.Net Core example demonstrating the basic use cases and how to set things up with this version of the SDK that will go into the https://github.com/growthbook/examples repo once this PR is complete and out to NuGet. 

I'm thinking that once the example PR is approved and merged over there then I can link it in the readme over here, would that be sufficient or do you think that it would still help to put some basic 'getting started' info here as well?

pascal89 commented:
I'm glad you've worked on some examples as well 😄 As long as we'll link to the example(s) from the readme it's fine for me. If you want my opinion on the examples you'll be adding, feel free to request me as a reviewer. I think that accounts for @Mastermindzh as well.
For this PR, IMO It's ready to go.

darklordzw commented:
PR has been merged into main. Thanks again for your contributions @Norhaven! I'll have a version up on NuGet in a few.
